### ISSUE TITLE
Misconfigured Google login leads to confusing redirect to admin settings; needs better explanation

### ISSUE THREAD

**Issue by paulproteus** [sandstorm-io/sandstorm#1355]
Steps to reproduce:

- Install Sandstorm as a self-hoster
- Enable Google login
- Log in
- Install Etherpad
- Create an Etherpad grain
- Create a sharing link
- Visit that sharing link in an incognito window
- Click on "Sign in" in the dropdown
- Click on the "Sign in with Google" choice

Expected behavior:

- See a login screen

Actual behavior:

- Find oneself at https://rose.sandcats.io/admin/settings which shows a message, "You are not logged in as admin and there isn't a valid token specified"

![anim](https://cloud.githubusercontent.com/assets/25457/12221348/341407ac-b74c-11e5-805d-ead8e0b1a270.gif)

**Comment by paulproteus** [comment]
Bizarrely, I can't reproduce this on oasis.sandstorm.io.

Sample sharing links that do repro the problem:

- https://rose.sandcats.io/shared/QtmEUkQGcgV2e3b0LSy6Qcq2-CrUqg8OYO8222cWtRd

- https://xamar.sandcats.io/shared/Bqa9dftNbc1Ni06D-SgBdkFuM_iky8VHAlTw0Rk1lzN (this is how I noticed the bug, thanks Pablo Jamar)

Sample sharing link that does not repro the problem:

- https://oasis.sandstorm.io/shared/MBt18sGiLbNG7UcexoHFvTbtcYEI8Uja_pzIVCMaVu0

**Comment by paulproteus** [comment]
Actually this seems to be a showstopper bug in login entirely. Steps to reproduce:

- Visit https://xamar.sandcats.io/

- Click "with Google"

Expected behavior:

- Be prompted for Google Account

Actual behavior:

- Find oneself at https://xamar.sandcats.io/admin/settings

**Comment by dwrensha** [comment]
I would guess that [this redirect](https://github.com/sandstorm-io/sandstorm/blob/3a847a42721305daf7931adf1007c0d5d4360e2e/shell/packages/sandstorm-accounts-ui/login_buttons.js#L99-L101) is getting tripped, which ought to only happen in the case of misconfiguration. I haven't been able to reproduce locally, so I'm not sure exactly what kind of misconfiguration that could be.

**Comment by kentonv** [comment]
Cannot reproduce on any of my servers, but a friend of @ndarilek reported this problem recently. He later found that his Google login configuration did not exist on the Google end. Upon reconfiguring it, the problem was fixed.

We should redirect to a page that says something like: "Google login is misconfigured on this server; please ask the administrator to reconfigure it."

**Comment by paulproteus** [comment]
But you can reproduce it at the two links I gave, on the two servers I
linked-to, right?

FWIW I don't think that rose.sandcats.io is misconfigured; I'm logged in to
it right now with a Google Account.

Mostly AFK today but can look into this in a few hours, else tomorrow.

**Comment by kentonv** [comment]
@paulproteus On both your server and xamar, Accounts.loginServiceConfiguration is empty:

    > Accounts.loginServiceConfiguration.find().fetch()
    []

On other servers, this collection is populated with OAuth configuration.

**Comment by kentonv** [comment]
Note that we have code which clears `loginServiceConfiguration` if `BASE_URL` is detected to have changed.

**Comment by paulproteus** [comment]
Hmm. I vaguely suspect this is some weird migration issue, because I have
not changed BASE_URL. But I can dig into this more by reading my logs
tomorrow.

**Comment by paulproteus** [comment]
Interesting!​ I confirm that my OAuth details are empty - but the OAuth
provider is still enabled.

1. How did this happen?

2. Probably Sandstorm should detect this weird case and show something
useful to the self-hosting admin so they know to fix it.

**Comment by dwrensha** [comment]
@paulproteus: Does the line "resetting oauth" appear recently in your sandstorm.log?

**Comment by dwrensha** [comment]
(The reset happens [here](https://github.com/sandstorm-io/sandstorm/blob/aef507ed7d13eeadf2740380781d40d87ae48ea6/shell/server/startup.js#L28-L32).)

**Comment by paulproteus** [comment]
Yes, I suppose it does.

```
** Starting Sandstorm at: Tue Jan  5 21:28:42 2016
** Starting back-end...
** Starting MongoDB...
note: noprealloc may hurt performance in many applications
about to fork child process, waiting until server is ready for connections.
forked process: 7
child process started successfully, parent exiting
** Back-end and Mongo started; now starting front-end...
Migrations already applied: 17/17
resetting oauth
** Shutting down due to signal
** Exiting
** Starting Sandstorm at: Tue Jan  5 21:29:48 2016
** Starting back-end...
** Starting MongoDB...
note: noprealloc may hurt performance in many applications
about to fork child process, waiting until server is ready for connections.
forked process: 7
child process started successfully, parent exiting
** Back-end and Mongo started; now starting front-end...
Migrations already applied: 17/17
resetting oauth
** Shutting down due to signal
** Exiting
** Starting Sandstorm at: Tue Jan  5 21:30:01 2016
** Starting back-end...
** Starting MongoDB...
note: noprealloc may hurt performance in many applications
```

Probably, while debugging things for other people on the mailing list and
testig out re-installs etc, I accidentally fiddled with the settings that
resulted in a removal of my OAuth configuration.

I guess from my perspective, nothing notified me in the UI that things were
broken. So maybe the simple fix here is that when resetting OAuth, we
should trigger a notification for all admin users.

**Comment by paulproteus** [comment]
OK, here are some cases that I think we should cover.

## We should show a note to the user in these cases

- The remote service tells us that its redirect_url etc. doesn't match what we have

- In our database, `Accounts.identityServices['google'].isEnabled()` is true but there is no `client_id`

## We should show a note to the admin in these cases

- When `BASE_URL` changes, resulting in `Accounts.loginServiceConfiguration` being removed

## Other note

- Perhaps when removing the `loginServiceConfiguration`, we should also make sure that the the provider returns false from e.g. `Accounts.identityServices['google'].isEnabled()` ?

**Comment by paulproteus** [comment]
Closed by #1387 

Many thanks to @dwrensha
